QoL Update: Improve Stability & Improvements

AdrianCassar opened the pull request:
- Games now working:
    - MotoGP 14
    - MotoGP 15
    - L4D
    - L4D GOTY
    - Team Fortress 2 TU 5

- Bugs Fixed
  - Create a session when STATS are enabled. (Fixes STATS for CSGO)
  - QoS (Quality of Service) data spamming the xenia web server causing timeouts from Heroku.
  - Issue with infinite thread spawning
  - Pointer in friends enumerator (Fixed Minecraft)
  - A bug with UPnP causing a crash.
  - Several null pointers.
  - A bug with IP URLs in api_address not adding `/`
  - FM4 `protect_zero = true` requirement.
  - TF2 `protect_zero = true` requirement.
  - RDR GOTY from crashing.

- Implemented
  - NetDll_XNetInAddrToString (Display IP in Aurora)
  - NetDll_XNetInAddrToServer
  - XNetLogonGetMachineID
  - SessionJoinLocal and SessionLeaveLocal.
  - SessionSearchByID

- XAM Stubs
  - XamQueryLiveHiveA (Fixed Minecraft crashing on later TUs)
  - XHttpOpenRequest (Fixed another Minecraft crash)

- General Improvements
  - XLiveAPI Initialization
  - Refactor SessionJoin and SessionLeave.
  - XUID Online/Offline Generation
  - If XUID is set to an empty string generate a new XUID
  - Netplay toolbar tab
  - Cleanup NetDll_XNetRandom
  - Add support for routers with permanent leases (UPnP)
  - Updated kernel version to latest [2.0.17559.0](https://support.xbox.com/en-GB/help/xbox-360/console/system-update-operating-system)
  - Updated readme with latest known working games.

- Regression
  - MotoGP 06 infinite loading screen when creating a session.
  - MW2 private match stuck joining session.
  - Leaderboard in L4D2 crashing due to CreateFriendsEnumerator.

- Config
  - `offline_mode` - Offline Mode
  - `log_mask_ips` - Do not include P2P IPs in xenia.log
  -  `stub_xlivebase` - Return success for all unimplemented XLiveBase calls

- XLiveBase
  - CreateFriendsEnumerator
  - XStorageBuildServerPath partial support
  - XStringVerify stub
  - DownloadToMemory stub
  - UploadFromMemory stub
  - XContentGetMarketplaceCounts stub
  - XUserMuteListQuery used by **MotoGP 15**
  - UploadFromMemory used by **Ultimate Marvel vs. Capcom 3**
  - 0x0005000D used by **Juiced 2: Hot Import Nights**

AdrianCassar committed:
Fixed QoS spam, improved XUIDs and bugs fixes

AdrianCassar committed:
[XAM] Implemented XNetInAddrToString & Updated kernel version

AdrianCassar committed:
[XAM] Create sessions for STAT enabled sessions and Stub XamQueryLiveHiveA

AdrianCassar committed:
[Xlivebase] Fixed pointer in friends enumerator

Gliniak commented on src/xenia/kernel/xam/xam_net.cc:
```diff
@@ -797,9 +831,13 @@ dword_result_t NetDll_XNetGetBroadcastVersionStatus_entry(dword_t caller,
-         XEthernetStatus::XNET_ETHERNET_LINK_FULL_DUPLEX;
+  if (cvars::offline_mode) {
+    return 0;
+  } else {
```
redundant else

AdrianCassar committed:
[XLiveAPI/APP] Improved XLiveAPI initialisation & Netplay toolbar tab

Improved XLiveAPI initialisation
Added Netplay toolbar tab

AdrianCassar committed:
[XAM] Fixed session creation flags

AdrianCassar committed:
[Xlivebase] Fixed GetServiceInfo

AdrianCassar committed:
[XAM] Fixed session creation flags & friends_count

AdrianCassar committed:
[XAM] Fixed XLiveAPI initialisation race condition

AdrianCassar committed:
[XAM] Implemented XNetLogonGetMachineID

Gliniak committed:
Sort messages in xlivebase (#6)

- Added description to some of them
- Added Empty function to add support for XStringVerify

imaginebeingatcomputers commented on README.md:
````diff
@@ -52,45 +52,60 @@ Where can I **download** the Canary Netplay build?
+Binding to privileged ports 1024 and below will usually fail on linux to fix this run this command.
+
+```console
+sudo sysctl net.ipv4.ip_unprivileged_port_start=0 
````
it should be noted that this will break ssh servers. I'd recommend instead something like 
`sudo sysctl net.ipv4.ip_unprivileged_port_start=999`

this should handle most cases (ports 1000-1002 which fall within Linux's default privileged range of 0-1024).

It should probably also be noted that simply running this will only fix the current session. sysctl settings will not be preserved after a reboot unless a config file is created. this can be done like so:
`echo 'sysctl net.ipv4.ip_unprivileged_port_start=999' | sudo tee -a /etc/sysctl.d/99-xenia.conf`

This will fix permanently fix most Linux issues around port binding, however if the user is on a Steam Deck, they will have no choice but to run the following on every boot.
`sudo sysctl net.ipv4.ip_unprivileged_port_start=999`

Altogether I think the section should read something like so:

## Linux Notes

### Failure to Bind to Ports
Binding to ports <= 1024 will usually fail on Linux as they are protected by default. To verify this is an issue you are encountering, search your log for the following message:
`NetDll_WSAGetLastError: 10013`

To fix this run this command:
```console
sudo sysctl net.ipv4.ip_unprivileged_port_start=999
echo 'sysctl net.ipv4.ip_unprivileged_port_start=999' | sudo tee /etc/sysctl.d/99-xenia.conf
```
This command configures privileged ports to start at port 999 instead of port 1024 in this logon session and future logons. This should allow for most games to now bind. 

If you are still seeing `NetDll_WSAGetLastError: 10013` in logs after running this, you can try rerunning the previous commands with a number lower than `999`. `23` should solve every case. You can try `0` but it will prevent you from running ssh.

It should also be noted that due to the way Steam Decks handle configuration, you will need to rerun this command on every reboot.

imaginebeingatcomputers reviewed (commented):
Linux notes

AdrianCassar committed:
[XAM] Fixed L4D & Refactor SessionJoin

- L4D & L4D GOTY now working.

- Fixed FM4 and RDR GOTY from crashing.

- Refactor SessionJoinRemote and SessionLeaveRemote.
- Implement SessionJoinLocal and SessionLeaveLocal for local users.

- Cleanup NetDll_XNetRandom

Gliniak committed:
[Xlivebase] Added partial support for message: 0x58035 - XStorageBuildServerPath (#7)

Gliniak committed:
[XAM] XGISessionJoin/Leave: Fixed incorrect pointer for local join (#8)

AdrianCassar committed:
[xlivebase] DownloadToMemory and UploadFromMemory stubs

AdrianCassar committed:
[UPnP] Add support for routers with permanent leases

AdrianCassar committed:
[UPnP] Cache whether router supports leases

AdrianCassar committed:
[UPnP] Add port results to Netplay Status tab

AdrianCassar committed:
[xlivebase] XContentGetMarketplaceCounts stub

Gliniak committed:
Removed hardcoded properties_count. Cleanup and simplification (#9)

AdrianCassar committed:
[XLiveAPI] Update DeleteSessions endpoint

AdrianCassar committed:
[XAM] Updated XNetGetTitleXnAddr

AdrianCassar committed:
[XAM] XHttpOpenRequest stub

Fixed a crash with Minecraft on latest TU

AdrianCassar committed:
[XLiveAPI] Added DeleteSessionsByMac endpoint

Delete sessions by mac when host or guest crashes.

Gliniak committed:
[Xsocket] Fixed issue with infinite thread spawning (#10)

AdrianCassar committed:
[UPnP] Fixed error type

AdrianCassar committed:
[XAM] Implemented XSessionSearchByID
